language/move-vm/runtime/src/loader.rs

```diff
@@ -2339,6 +2354,15 @@ impl Loader {
         })
     }
 
+    fn count_type_nodes(&self, ty: &Type) -> usize {
+        match ty {
+            Type::Vector(ty) => self.count_type_nodes(&ty).saturating_add(1),
+            Type::StructInstantiation(_, ty_args) => ty_args.iter().fold(1, |x, y| x.saturating_add(self.count_type_nodes(y))),
+            Type::Reference(t) | Type::MutableReference(t) => self.count_type_nodes(t).saturating_add(1),
+            _ => 1
+        }
+    }
```

Comment:
This is an unbounded recursion that might be limited by other restrictions but seems risky. Instead could we pass in the value up to this point (e.g., tail recursion of sorts) and terminate if it exceeds a maxium?

Reply:
Made this iterative

Reply:
For the record, I think the termination check is unnecessary since the point here is that those types which lead to overflow can never be constructed. We already have multiple recursive algorithms which do no checks at all and expect the thing is sized in range. This is just one more. Its purpose is to disallow constructing types which are too large.

Reply:
Yeah we have some other checks ensuring that VM does not create a type that's over 256 levels deep so this shouldn't be a problem, at least for now